include/swift/Runtime/MutexPThread.h

```diff
@@ -0,0 +1,118 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Mutex, Condition, Read/Write lock, and Scoped lock implementations
+// using PThreads.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_RUNTIME_MUTEX_PHTREAD_H
+#define SWIFT_RUNTIME_MUTEX_PHTREAD_H
+
+#include <pthread.h>
+
+namespace swift {
+
+/// Private PThread implementation of Condition functionality.
+/// Use Condition instead (see Mutex.h).
+class ConditionImpl {
+  friend class MutexImpl;
+  friend class Condition;
+
+  ConditionImpl(const ConditionImpl &) = delete;
+  ConditionImpl &operator=(const ConditionImpl &) = delete;
+  ConditionImpl(ConditionImpl &&) = delete;
+  ConditionImpl &operator=(ConditionImpl &&) = delete;
+
+public:
+  ~ConditionImpl();
+
+public:
+  void notifyOne();
+  void notifyAll();
+
+private:
+  explicit ConditionImpl();
+  pthread_cond_t PThreadCond;
+};
+
+/// Private PThread implementation of Mutex functionality.
+/// Use Mutex instead (see Mutex.h).
+class MutexImpl {
+  friend class Mutex;
+
+  MutexImpl() = delete;
+  MutexImpl(const MutexImpl &) = delete;
+  MutexImpl &operator=(const MutexImpl &) = delete;
+  MutexImpl(MutexImpl &&) = delete;
+  MutexImpl &operator=(MutexImpl &&) = delete;
+
+public:
+  ~MutexImpl();
+
+public:
+  void lock();
+  void unlock();
+  bool try_lock();
+  void wait(ConditionImpl &condition);
+
+private:
+  explicit MutexImpl(bool checked);
+  pthread_mutex_t PThreadMutex;
+};
+
+/// Private PThread implementation of ReadWriteLock functionality.
+/// Use ReadWriteLock instead (see Mutex.h).
+class ReadWriteLockImpl {
+  friend class ReadWriteLock;
+
+  ReadWriteLockImpl(const ReadWriteLockImpl &) = delete;
+  ReadWriteLockImpl &operator=(const ReadWriteLockImpl &) = delete;
+  ReadWriteLockImpl(ReadWriteLockImpl &&) = delete;
+  ReadWriteLockImpl &operator=(ReadWriteLockImpl &&) = delete;
+
+public:
+  ~ReadWriteLockImpl();
+
+public:
+  void readLock();
+  bool try_readLock();
+  void writeLock();
+  bool try_writeLock();
+  void unlock();
+
+private:
+  explicit ReadWriteLockImpl();
+  pthread_rwlock_t PTheadRWLock;
+};
+
+/// Use UnsafeMutex instead (see Mutex.h).
+class UnsafeMutexImpl {
+  friend class UnsafeMutex;
+
+  UnsafeMutexImpl(const UnsafeMutexImpl &) = delete;
+  UnsafeMutexImpl &operator=(const UnsafeMutexImpl &) = delete;
+  UnsafeMutexImpl(UnsafeMutexImpl &&) = delete;
+  UnsafeMutexImpl &operator=(UnsafeMutexImpl &&) = delete;
+
+public:
+  ~UnsafeMutexImpl() { pthread_mutex_destroy(&PThreadMutex); }
```

Comment:
Yeah global destructors are evil in my experience, dealing with this by providing static versions as needed.